engineering/adr-0002-height-and-round-in-unlock-tx-payload.md

```diff
@@ -0,0 +1,101 @@
+
+```
+requestID = SHA256("dpbvote", height, round)
+
+sigHash = SHA256(transaction without quorumSig)
+
+SignID = SHA256(quorumHash, requestID, sigHash)
+```
+
+Although this approach protects against the attack, there is no way for consumers to verify this signature
+since they don't have platform height and round.
+
+On other hand, Drive implements a special two block commit approach to construct and sign asset unlock transactions in safe way.
+With this solution the transactions aren't signed immediately but incomplete form of them (without block specific information in payload)
+go to special queue and will be pulled for signing on the next block. This guarantee on code level that for specific height only those
+exact transactions will be used as vote extensions for every round. There are three types of asset unlock transactions, additional queue
+in platform state and logic in Drive implemented to support this solution.
+
+There two ways to proceed with credits withdrawal implementation:
+1. Add height and round to asset lock transaction payload and remove unnecessary in this case "two block commit" solution to simplify the system
+2. Change tenderdash to sign the whole transaction instead of DIP7 format.
+
+Both approaches have benefits and downsides. In this ADR we will look closer to the first solution.
+There is [ADR0003](./adr-0003-vote-extensions-without-DIP7.md) that is dedicated to the second approach.
+
+## Decision
+
+To allow asset unlock transaction consumers verify signature, concatenated prefix `ve` (Dash Platform Vote Extension), platform height (uint64)
+and platform round (uint32) as `signatureRequest` (14 bytes) field should be added to asset unlock transaction payload. Having a complete request
+message as part of the transaction payload will allow not just tenderdash but other L2 systems to utilize Asset Lock/Unlock Core functionality for
+their needs in the future.
+
+To verify the signature, asset unlock transaction consumer should hash `signatureRequest` from payload a combine with `quorumHash` and `sigHash`
+according to DIP7:
+
+```
+requestID = SHA256(signatureRequest from transaction payload)
+
+sigHash = SHA256(transaction without quorumSig)
+
+SignID = SHA256(quorumHash, requestID, sigHash)
+```
+
+The two block commit logic that mitigates "multi round signature shares" attack using additional step with a queue become unnecessary and can be removed from the codebase to simplify the withdrawal process.
+
+An additional security layer could be added to [Sign](https://dashcore.readme.io/docs/core-api-ref-remote-procedure-calls-evo#quorum-sign) Core's RPC.
+List of quorums types should be limited to platform only. It will guarantiee that Sign RPC method won't be used to attack Layer 1 functionality like
+Instant Send and Chain Lock. 
+
+
+## Implementation
```

Comment:
Having round in transaction make them different between the rounds. So if first round fails but an attacker could still collect shares and sign with his part he can broadcast this transaction to the Core and get money from nothing.

Reply:
Great finding!

It's not an issue from Tenderdash/Drive perspective, as these components are aware of committed round and can verify this info. But Core doesn't have the commit, right?

On the other hand, we can't remove it, otherwise we can collect vote extensions from multiple rounds and recover signature.